### PR TITLE
feat(eslint-plugin): Add rule no-reference-import

### PR DESCRIPTION
**Pull request by jessetrinity**
Adding rule for corresponding tslint rule [no-reference-import](https://palantir.github.io/tslint/rules/no-reference-import/).

Should disallow doing both

```ts
/// <reference types="foo" />
//and
import * as foo from "foo"
```

in the same file.

Some thoughts:

1. I wasn't sure if it would be better to gate the rule checking behind the presence of the triple slash directive - input there would be appreciated.
2. How does one determine what the recommendation level / severity of a rule should be?
3. I probably need more test cases - I put in the obvious ones, but I could be more familiar with the module system.

### CODE REVIEW

**Comment by bradzacher** [comment]
> How does one determine what the recommendation level / severity of a rule should be?

Set it to false (i.e. not recommended).
We cannot make changes to the recommended set without releasing a breaking change.


---

What are the situations that you might want triple slash reference imports?
I'm just wondering what the use cases are for this rule when compared to `no-triple-slash-reference`

**Comment by uniqueiniquity** [comment]
@bradzacher You'd want triple-slash references if you're writing an app without using modules but still want to enforce a dependency order between your source files. The TypeScript codebase is actually written like this - the whole thing is written with merged namespaces and triple-slash references that compile to a single output file.

File: packages/eslint-plugin/src/rules/no-reference-import.ts [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+            data: {
+              module: referenceResult[1],
+            },
+          });
+        }
+      });
+    }
+    return {
+      ImportDeclaration(node) {
+        if (node.type === AST_NODE_TYPES.ImportDeclaration) {
```
Comment (uniqueiniquity) [inline packages/eslint-plugin/src/rules/no-reference-import.ts]: I think this is always true, so isn't necessary. Same on line 60.

File: packages/eslint-plugin/src/rules/no-reference-import.ts [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    function hasMatchingReference(source: Literal) {
+      const referenceRegExp = /^\/\s*<reference\s*types="(.*)"/;
+      const sourceCode = context.getSourceCode();
+      const commentsBefore = sourceCode.getCommentsBefore(programNode);
+
+      commentsBefore.forEach(comment => {
+        if (comment.type !== 'Line') {
+          return;
+        }
+        const referenceResult = referenceRegExp.exec(comment.value);
```
Comment (uniqueiniquity) [inline packages/eslint-plugin/src/rules/no-reference-import.ts]: It seems like you're running this every time you see an import node.
Could you run this once and extract the list of referenced packages from the references, then look up in that list on each import node?

**Comment by uniqueiniquity** [comment]
@bradzacher Let's try this again.
This rule addresses a different kind of triple-slash-reference than `no-triple-slash-reference`.
I think the use case here would be if you wanted access to types from a node_modules folder, but didn't want to make your file a module.

**Comment by JamesHenry** [comment]
A good case is needing to directly tie a spec/test file to the types for the test runner you are using.

The majority of test runners just create globals for their utils, and, particularly in a monorepo, you may have multiple different test runners doing this, making it difficult to have the right ambient types associated with the right files.

I believe I also remember Alex Eagle mentioning this is also a consideration for JS test runners in a Bazel context, where Bazel has to know all the inputs explicitly in order to function correctly.

File: packages/eslint-plugin/src/rules/no-reference-import.ts (hunk 2 above)
Comment (jessetrinity) [reply packages/eslint-plugin/src/rules/no-reference-import.ts]: Moved the collecting references bit to the program node.

**Comment by bradzacher** [comment]
sorry If I'm being a little slow here...
It looks like this rule builds off of same logic as `no-triple-slash-reference`?

`no-triple-slash-reference`:
https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/src/rules/no-triple-slash-reference.ts#L23-L37

`no-reference-import`:
https://github.com/typescript-eslint/typescript-eslint/blob/bc6316599e84a6e503d66a6ac85becc704063a0a/packages/eslint-plugin/src/rules/no-reference-import.ts#L60-L75

Couldn't the new logic be placed inside `no-triple-slash-reference` as an option instead of creating a new rule with duplicate logic?

**Comment by uniqueiniquity** [comment]
@bradzacher Sure, we could combine the two; they're separate at least because they were separate in TSLint. However, note that it's not as simple as an option to toggle between the two implementations; in a unified rule, you'd have to have an option for whether you wanted to allow or disallow `///<reference path=... />` and another option for what behavior you want for `///<reference types=... />`. It's certainly feasible to combine those, though, especially if there's any perf advantages that you're aware of (beyond not iterating through the comments at the top of the file twice, of course).

**Comment by jessetrinity** [comment]
@bradzacher in addition to what @uniqueiniquity said, this rule does not strictly ban `/// <reference types="" />` directives like the other rule does, but rather prevents you from using them if you also use an import declaration for that package.

A unified rule would need to specify severity levels for each species of reference directive, something like `no-duplicate` and `banned` for each of `/// <reference path="" />` and `/// <reference types="" />` (we also have `/// <reference lib="" />`).

I don't know the logic behind tslint's decision to have a rule banning one and de-duping the other. I have no problem combining them but I think we were converting rules in the spirit of being consistent with tslint.

**Comment by bradzacher** [comment]
AHH! That's what I missed:  `/// <reference ****types****=....`. SORRY.

---

Spitballing here.

What do you guys think about this direction:
- rename this new rule to `triple-slash-reference`
    - we're kinda trying to remove `no-*` rule names.
- add an options object (see below)
- deprecate `no-triple-slash-reference` naming this new rule.

```ts
type Options = [
    {
        lib: 'never' | 'allow' | 'allow-when-not-imported',
        path: 'never' | 'allow',
        ref: 'never' | 'allow',
    }
];
```

**Comment by uniqueiniquity** [comment]
Seems reasonable to me. I assume you meant `lib`, `path`, and `types`, but let me know if I'm missing something.

Thoughts, @jessetrinity?

**Comment by jessetrinity** [comment]
Sure.

> ```ts
> lib: 'never' | 'allow' | 'allow-when-not-imported',
> ```

Is there already a pattern for rule option names? Something like `'banned' | 'allow' | 'prefer-import'` would be more obvious to me.

**Comment by bradzacher** [comment]
nah there's no guidelines, it's more of a "whatever makes sense" thing.
Maybe `'ban'` instead of `'banned'`. Just to keep the tense consistent with `'allow'`.
everything else looks good! thanks!

**Comment by jessetrinity** [comment]
Maybe I'm being too conservative, but I did not add the `prefer-import` option to `/// <reference path="" />` or `/// <reference lib="" />` directives as they were not covered in existing rules.

File: packages/eslint-plugin/src/rules/no-triple-slash-reference.ts [hunk 3]
```diff
@@ -11,7 +11,7 @@ export default util.createRule({
     },
     schema: [],
     messages: {
-      tripleSlashReference: 'Do not use a triple slash reference.',
+      noTripleSlashReference: 'Do not use a triple slash reference.',
     },
```
Comment (bradzacher) [inline packages/eslint-plugin/src/rules/no-triple-slash-reference.ts]: add `deprecated: true`, and `replacedBy: ['triple-slash-reference']` to the meta

**Comment by JamesHenry** [comment]
@bradzacher it seems to be failing on documentation validation, any tips for @jessetrinity on how to get the build to pass?

**Comment by bradzacher** [comment]
glancing at the build output
https://dev.azure.com/typescript-eslint/TypeScript%20ESLint/_build/results?buildId=1344

`Found rule no-triple-slash-reference in table, but it doesn't exist in the plugin.`
Just need to delete the `no-triple-slash-reference` line from the table in the readme (the rule has been deprecated, so it shouldn't be advertised in the root readme any more)

**Comment by milesj** [comment]
This was super confusing. The title says its `no-reference-import`, while the old tsline was `no-triple-slash-reference`, but now its `triple-slash-reference`.

**Comment by jessetrinity** [comment]
> This was super confusing. The title says its `no-reference-import`, while the old tslint was `no-triple-slash-reference`, but now its `triple-slash-reference`.

Sorry about that. There is a separate tslint `no-reference-import` rule this was meant to mirror. it ended up absorbing `no-triple-slash-reference` so we tried a more generic name. I think there is a desire to move away from the "no-foo" naming convention.

**Comment by bradzacher** [comment]
sorry! I've updated the release notes / pr title.

@jessetrinity is correct. The standard we want to move toward is not having `no-*` rules unless it makes literally no sense to ever use the `*`.
Additionally tslint had two rules which did very similar things. Because we're doing a clean implementation, it made sense to merge the rules together whilst making it more flexible by allowing the inverse.

The old rule `no-triple-slash-reference` still exists in v1 of the plugin, but will be removed in v2.

**Comment by milesj** [comment]
Makes sense, thanks for the quick turn around!

**Pull request merged**